src/Servers/HttpSys/test/FunctionalTests/ResponseTrailersTests.cs

```diff
@@ -0,0 +1,384 @@
+        public async Task ResponseTrailers_ProhibitedTrailers_Blocked()
+        {
+            using (Utilities.CreateDynamicHttpsServer(out var address, httpContext =>
+            {
+                Assert.True(httpContext.Response.SupportsTrailers());
+                foreach (var header in _disallowedTrailers)
+                {
+                    Assert.Throws<InvalidOperationException>(() => httpContext.Response.AppendTrailer(header, "value"));
+                }
+                return Task.FromResult(0);
+            }))
+            {
+                var response = await SendRequestAsync(address);
+                response.EnsureSuccessStatusCode();
+                Assert.Equal(HttpVersion.Version20, response.Version);
+                Assert.Empty(response.TrailingHeaders);
+            }
+        }
+
+        [ConditionalFact]
+        [MinimumOSVersion(OperatingSystems.Windows, "10.0.19505", SkipReason = "Requires HTTP/2 Trailers support.")]
+        public async Task ResponseTrailers_NoBody_TrailersSent()
+        {
+            using (Utilities.CreateDynamicHttpsServer(out var address, httpContext =>
+            {
+                httpContext.Response.DeclareTrailer("trailername");
+                httpContext.Response.AppendTrailer("trailername", "TrailerValue");
+                return Task.FromResult(0);
+            }))
+            {
+                var response = await SendRequestAsync(address);
+                response.EnsureSuccessStatusCode();
+                Assert.Equal(HttpVersion.Version20, response.Version);
+                Assert.NotEmpty(response.TrailingHeaders);
+                Assert.Equal("TrailerValue", response.TrailingHeaders.GetValues("TrailerName").Single());
+            }
+        }
+
+        [ConditionalFact]
+        [MinimumOSVersion(OperatingSystems.Windows, "10.0.19505", SkipReason = "Requires HTTP/2 Trailers support.")]
+        public async Task ResponseTrailers_WithBody_TrailersSent()
+        {
+            using (Utilities.CreateDynamicHttpsServer(out var address, async httpContext =>
+            {
+                await httpContext.Response.WriteAsync("Hello World");
+                httpContext.Response.AppendTrailer("TrailerName", "Trailer Value");
+            }))
+            {
+                var response = await SendRequestAsync(address);
+                response.EnsureSuccessStatusCode();
+                Assert.Equal(HttpVersion.Version20, response.Version);
+                Assert.Equal("Hello World", await response.Content.ReadAsStringAsync());
+                Assert.NotEmpty(response.TrailingHeaders);
+                Assert.Equal("Trailer Value", response.TrailingHeaders.GetValues("TrailerName").Single());
+            }
+        }
+
+        [ConditionalFact]
+        [MinimumOSVersion(OperatingSystems.Windows, "10.0.19505", SkipReason = "Requires HTTP/2 Trailers support.")]
+        public async Task ResponseTrailers_WithContentLengthBody_TrailersNotSent()
```

Comment:
Is this consistent with Kestrel, or different? (note: gRPC doesn't use ContentLength so not important to us, just curious)

Reply:
It's different. We've talked for a long time about having Kestrel end responses as soon as the last byte was written for reduced latency, but we knew it would interfere with trailers. This is the first time we've been forced to reconcile the two features.